docs: add i18n overview & authoring documentation (#9361)
GoogleChrome/lighthouse: lighthouse-core/lib/i18n/README.md

```

Note: Why do `en-US` and `en-XL` get baked early?  We write all our strings in `en-US` by default, so they do not need to be translated, so it can be immediately baked without going to the translators.  Similarly, `en-XL` is a debugging language, it is an automated version of `en-US` that simply adds markers to `en` strings in order to make it obvious that something has or hasn't been translated.  So neither of these files need to go to translators to be used, and both can be used at develop-time to help developer i18n workflow.

#### String Translation in Google Translation Console

* Googler is ready to kick off the TC pipeline again.

```shell
# collect UIStrings (to make sure everything is up to date)
$ yarn i18n

# Extract the CTC format files to translation console
$ sh import-source-from-github.sh

# Submit CL. Wait ~2 weeks for translations

# Import the translated CTC format files to locales/ and bake them
$ sh export-tc-dump-to-github.sh
```

#### String Replacement (runtime)

See [Appendix A: How runtime string replacement works](#appendix)

# Writing UIStrings with LHL

> ❗TODO(exterkamp): explain all the comments and where they go/what they become.

Reviewer: should we leave this out/file issues for them?

Reviewer 2: imo I was going to handle that in #9114 once it was reimported :man_shrugging:

Reviewer 3: should we just block this PR on #9114? The implementation may still change until that lands...

Author: let's land the doc. 9114 can always adjust anything as need be.

Reviewer 3: I'd like to do a review then, please, so I'm not just doing a review of it in #9114 after all

Author: Ooops. Sorry just saw this right after i merged. Happy to take your review and i'll do whatever followups you're thinking

❗TODO(exterkamp): explain why we can't use some ICU like number formatting.

We want to keep strings close to the code in which they are used so that developers can easily understand their context. We use `i18n.js` to extract the `UIStrings` strings from individual js files.

LHL strings in each module are defined in a `UIStrings` object with the strings as its properties. JSDoc is sometimes used to provide additional information about each string.

The LHL syntax is based primarily around the standardized [ICU message formatting](http://userguide.icu-project.org/formatparse/messages) syntax.

### Basic example

A simple string.

```javascript
const UIStrings = {
  /** Imperative title of a Lighthouse audit that ... */
  title: 'Minify CSS',
};
```

For proper translation, **all** strings must be accompanied by a description, written as a preceeding comment.

### Replacements and primitive formatting

> Replacements (aka substitutions) include string replacements like `{some_name}` number formatting like `{timeInMs, number, milliseconds}`.

Reviewer: ```suggestion
Replacements (aka substitutions) include string replacements like `{some_name}` and number formatting like `{timeInMs, number, milliseconds}`.
```

#### Direct ICU replacement

`{some_name}` is called _Direct ICU_ since the replacement is a direct substitution of ICU with a variable and uses no custom formatting. This is simply a direct replacement of text into a string. Often used for proper nouns, code, or other text that is dynamic and added at runtime.

ICU replacements must use a JSDoc-like syntax to specify an example for direct ICU replacements:

*   To specify the description, use
    * `@description <description>`
*   To specify an example for an ICU replacement, use

>     * `@example {<ICU variable name>} <example for ICU replacement>`

Reviewer: are we leaving it at `{variableName} replacement example` then?

```javascript
const UIStrings = {
    /**
      * @description Error message explaining ...
      * @example {errorCode} NO_SPEEDLINE_FRAMES
      */
    didntCollectScreenshots: `Chrome didn't .... ({errorCode})`,
};
```

#### Complex ICU replacement

> `{timeInMs, number, milliseconds}` is called _Complex ICU_ since the replacement is for complex numbers and uses the custom formatters in Lighthouse. The supported complex ICU formats are: `milliseconds`, `seconds`, `bytes`, `percent`, and `extendedPercent`.

Reviewer: ```suggestion
`{timeInMs, number, milliseconds}` is called _Complex ICU_ since the replacement is for numbers and other complex replacements that use the custom formatters in Lighthouse. The supported complex ICU formats are: `milliseconds`, `seconds`, `bytes`, `percent`, and `extendedPercent`.
```

Reviewer: not attached to the specific language but `complex numbers` is it's own thing so would prefer to just leave it at complex replacements ;)

> These complex ICU formats are automatically given @example values during `yarn i18n`.  Therefore, a normal description string can be used:

Reviewer: ❤️ dope!

```javascript
const UIStrings = {
    /** Description of display value. */
    displayValueText: 'Interactive at {timeInMs, number, seconds} s',
};
```

### Ordinals (Numeric Selects), Plurals

An ordinal ICU message is used when the message contains "plurals", wherein a sub-message would need to be selected from a list of messages depending on the value of `itemCount` (in this example).  They are a flavor of "Selects" that have a unique syntax.

```javascript
displayValue: `{itemCount, plural,
  =1 {1 link found}
  other {# links found}
  }`,
```

### Selects

A select ICU message is used when the message should select a sub-message based on the value of a variable `pronoun` in this case. This is often used for gender based selections, but can be used for any enum.  Lighthouse does not use selects very often.

```javascript
displayValue: `{pronoun, select,
  male {He programmed the link.}
  female {She programmed the link.}

>   other {They programmed the link}

Reviewer: ```suggestion
  other {They programmed the link.}
```
